### PR TITLE
Add serverCertificate connection property for encrypt=strict

### PR DESCRIPTION
**Pull request by lilgreenbird**

### CODE REVIEW

**Review by Jeffery-Wasty — APPROVED** [review]

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java [hunk 1]
```diff
@@ -63,6 +73,233 @@
         }
     }
 
+    /**
+     * Parse name in RFC 2253 format Returns the common name if successful, null if failed to find the common name. The
+     * parser tuned to be safe than sorry so if it sees something it can't parse correctly it returns null
+     */
```
Comment (tkyc) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java:79]: Missing java doc annotations

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java [hunk 2]
```diff
@@ -906,4 +909,26 @@ public void testPersistentTokenCacheAccessAspect() throws SQLException {
...
+    public void testServerCert() throws SQLException {
+        SQLServerDataSource ds = new SQLServerDataSource();
+        ds.setURL(connectionString);
+        ds.setServerCertificate("badCert");
+        ds.setEncrypt(Constants.STRICT);
+        ds.setTrustServerCertificate(false);
+        try (Connection con = ds.getConnection()) {
+            fail(TestResource.getResource("R_expectedFailPassed"));
+        } catch (SQLException e) {
+            // TODO: servers which do not support TDSS will return SSL failed error, test should be updated once server
```
Comment (tkyc) [inline src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java:928]: Hmm, I was going to ask about more tests eg. maybe one test that tests the connection string, but without servers that support TDSS, it doesn't seem possible?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java [hunk 3]
```diff
@@ -63,6 +73,233 @@
...
+        return commonName;
+    }
+
+    /**
+     * Validate server name in certificate matches hostname
+     * 
+     * @param nameInCert
+     *        server name in certificate
+     * @param hostName
+     *        * hostname
```
Comment (v-zhangw) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java:117]: Why extra "*" here?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java [hunk 4]
```diff
@@ -63,6 +73,233 @@
...
+     * @return
```
Comment (v-zhangw) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java:118]: Missing description for @return

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java [hunk 5]
```diff
@@ -63,6 +73,233 @@
...
+                            // As per RFC2459, the DNSName will be in the
+                            // "preferred name syntax" as specified by RFC
+                            // 1034 and the name can be in upper or lower case.
+                            // And no significance is attached to case.
+                            // Java docs guarantee that the second entry in the list
+                            // will be a string for dnsName
+                            if (value != null && value instanceof String) {
+                                dnsNameInSANCert = (String) value;
+
+                                // Use English locale to avoid Turkish i issues.
```
Comment (v-zhangw) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java:227]: Potential typo? What are Turkish "i" issues?

File: src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java [hunk 6]
```diff
@@ -1895,88 +1626,109 @@ else if (con.getTrustManagerClass() != null) {
             // Otherwise, we'll validate the certificate using a real TrustManager obtained
             // from the a security provider that is capable of validating X.509 certificates.
             else {
-                if (logger.isLoggable(Level.FINER))
-                    logger.finer(toString() + " SSL handshake will validate server certificate");
+                if (isTDSS) {
+                    if (logger.isLoggable(Level.FINEST))
+                        logger.finest(toString() + " Verify server certificate for TDSS");
+
+                    if (logger.isLoggable(Level.FINEST))
```
Comment (v-zhangw) [inline src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java:1633]: 1630, 1633 and 1638 are all checking the same condition. All this logging can be moved under the first FINEST check.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerTrustManager.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+        defaultTrustManager.checkServerTrusted(chain, authType);
+
+        // Explicitly validate the expiry dates
+        for (X509Certificate cert : chain) {
+            cert.checkValidity();
+        }
+
+        if (logger.isLoggable(Level.FINEST)) {
+            logger.finest(logContext + " default serverTrusted succeeded proceeding with server name validation");
```
Comment (v-zhangw) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerTrustManager.java:95]: Nitpick: Capitalize D in default. Add comma after first clause (default serverTrusted succeeded).

**Review by tkyc — APPROVED** [review]

**Review by Jeffery-Wasty — APPROVED** [review]

**Review by v-zhangw — APPROVED** [review]

**Review by v-zhangw — APPROVED** [review]

**Pull request merged**